File: algorithms/easy/minimum_hours_of_training_to_win_a_competition.ex
defmodule Solution do
  @spec min_number_of_hours(
          initial_energy :: integer,
          initial_experience :: integer,
          energy :: [integer],
          experience :: [integer]
        ) :: integer
  def min_number_of_hours(initial_energy, initial_experience, energy, experience) do
    ex_training =
      experience
      |> Enum.reduce({initial_experience, 0}, fn ex, {init, acc} ->
        if init <= ex do
          {init + ex + ex - init + 1, acc + ex - init + 1}
        else
          {init + ex, acc}
        end
      end)
      |> elem(1)

    all_energy = Enum.sum(energy)

    en_training = if initial_energy > all_energy, do: 0, else: all_energy - initial_energy + 1
    ex_training + en_training
  end
end

File: algorithms/easy/minimum_hours_of_training_to_win_a_competition.cpp
class Solution {
public:
  int minNumberOfHours(int initialEnergy, int initialExperience, vector<int>& energy, vector<int>& experience) {
    int allEnergy = accumulate(energy.begin(), energy.end(), 0);
    int trainingEnergy = initialEnergy > allEnergy ? 0 : allEnergy - initialEnergy + 1;
    int trainingExp = 0;
    for (auto& ex : experience) {
      if (initialExperience <= ex) {
        int diff = ex - initialExperience + 1;
        initialExperience += diff;
        trainingExp += diff;
      }
      initialExperience += ex;
    }
    return trainingEnergy + trainingExp;
  }
};
